src/app/clusters/scenes/ExtensionFieldsSetsImpl.h

```diff
@@ -0,0 +1,159 @@
+    {
+        TLV::TLVType container;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::ContextTag(1), TLV::kTLVType_Structure, container));
+
+        ReturnErrorOnFailure(writer.Put(TagClusterId(), static_cast<uint16_t>(this->mID)));
+        ReturnErrorOnFailure(writer.PutBytes(TagEFS(), mBytesBuffer, mUsedBytes));
+
+        return writer.EndContainer(container);
+    }
+    CHIP_ERROR Deserialize(TLV::TLVReader & reader)
+    {
+        ByteSpan buffer;
+        TLV::TLVType container;
+        ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::ContextTag(1)));
+        ReturnErrorOnFailure(reader.EnterContainer(container));
+
+        ReturnErrorOnFailure(reader.Next(TagClusterId()));
+        ReturnErrorOnFailure(reader.Get(this->mID));
+
+        ReturnErrorOnFailure(reader.Next(TagEFS()));
+        ReturnErrorOnFailure(reader.Get(buffer));
+        if (buffer.size() > kMaxFieldsPerCluster)
+        {
+            this->mUsedBytes = kMaxFieldsPerCluster;
+        }
+        else
+        {
+            this->mUsedBytes = static_cast<uint8_t>(buffer.size());
+        }
+        memcpy(this->mBytesBuffer, buffer.data(), this->mUsedBytes);
+
+        return reader.ExitContainer(container);
+    }
+
+    void Clear()
+    {
+        this->mID = kInvalidClusterId;
+        memset(this->mBytesBuffer, 0, kMaxFieldsPerCluster);
+        this->mUsedBytes = 0;
+    }
+
+    bool is_empty() const { return (this->mUsedBytes == 0); }
+
+    bool operator==(const ExtensionFieldsSet & other)
+    {
+        return (this->mID == other.mID && !memcmp(this->mBytesBuffer, other.mBytesBuffer, this->mUsedBytes) &&
+                this->mUsedBytes == other.mUsedBytes);
+    }
+
+    CHIP_ERROR operator=(const ExtensionFieldsSet & other)
+    {
+        if (other.mUsedBytes <= kMaxFieldsPerCluster)
+        {
+            memcpy(this->mBytesBuffer, other.mBytesBuffer, other.mUsedBytes);
+        }
+        else
+        {
+            return CHIP_ERROR_BUFFER_TOO_SMALL;
+        }
+        this->mID        = other.mID;
+        this->mUsedBytes = other.mUsedBytes;
+
+        return CHIP_NO_ERROR;
+    }
+};
+
+class ExtensionFieldsSetsImpl : public ExtensionFieldsSets
+{
+public:
+    ExtensionFieldsSetsImpl();
+    ~ExtensionFieldsSetsImpl() override{};
+
+    // overrides
+    CHIP_ERROR Serialize(TLV::TLVWriter & writer) const override;
+    CHIP_ERROR Deserialize(TLV::TLVReader & reader) override;
+    void Clear() override;
+    bool is_empty() const override;
+
+    // implementation
+    CHIP_ERROR insertField(ExtensionFieldsSet & field);
+    CHIP_ERROR getFieldAtPosition(ExtensionFieldsSet & field, uint8_t position);
+    CHIP_ERROR removeFieldAtPosition(uint8_t position);
+
+    bool operator==(const ExtensionFieldsSetsImpl & other)
+    {
+        for (uint8_t i = 0; i < kMaxClusterPerScenes; i++)
+        {
+            if (!(this->mEFS[i] == other.mEFS[i]))
+            {
+                return false;
+            }
```

Comment:
Note that after removals, there will be "empty" slots in the extension field set. Why does the implementation not "compress" this down?

Reply:
Implementation now compresses the array on removal.